### ISSUE TITLE
在SalienceCriterion类中，引入新的线性层参与计算，报错未收到梯度的问题（Parameters which did not receive grad for rank 0）

### ISSUE THREAD

**Issue by autoencoder-zz** [xiuqhou/Salience-DETR#42]
### Question

作者你好，再次打扰您一下，我在魔改模型的时候遇到了参数未收到梯度的问题

具体来说就是我定义了一个A类，其实就是一些线性层，在SalienceCriterion类中初始化A类，使其参与confidence部分的计算，结果运行之后报错Parameters which did not receive grad，在查看具体参数后，发现是

` focus_criterion.A.layer.4.bias,  focus_criterion.A.layer.4.weight, 
focus_criterion.A.layer.2.bias,   focus_criterion.A.layer.2.weight,   
focus_criterion.A.layer.0.bias,   focus_criterion.A.layer.0.weight`

我想让A类的权重和偏置参与反向传播不断学习更合适的值，但是我发现报错的这些是1、3、5的权重和偏置，我不明白为什么会这样。

希望您能抽时间为我讲解一下，谢谢！

**Comment by autoencoder-zz** [comment]
还有另外一个问题

关于配置中的800*1333

这个是min_size=800,    max_size=1333是如何确定的呢？

我想换别的数据集测试一下，想对最小最大值进行修改，以期更好的精度，谢谢。

**Comment by xiuqhou** [comment]
Hi @autoencoder-zz 感谢您的提问~

第一个问题是因为最终的loss没有用到您定义的这些层，请检查是否某些操作中断了这些层相关的梯度，或者这些层的输出压根就没有用到loss计算中，或者看日志有没有输出对应的loss。**确保这些层的参数经过层层前向传播后会参与到最终的loss中！**

第二个问题，800\*1333是常见的COCO训练设置，如果用您自己的数据集，参考大部分图片的宽高来设置就行了，没有什么特定的标准。例如您可以统计下图片h的平均值、w的平均值，根据他们来设置为min_size和max_size。

**要注意的是**，当您修改min_size和max_size的时候，务必也要相应修改`train_transform`中的图片尺寸，在`transforms/presets.py`中我定义了800\*1333对应的一些数据增强，当选择其他min_size和max_size的时候，也要按照比例来修改数据增强中的图片尺寸，改完后在train_config.py中使用。

**Comment by autoencoder-zz** [comment]
很不幸，我仔细排查了一下，并没有发现有未使用的返回值，改动、增加的变量都参与到计算当中了。

我定义的层是这样的
```
self.layer = nn.Sequential(
            nn.Linear(in_dim, in_dim // 2),
            nn.GELU(),
            nn.Linear(in_dim // 2, in_dim // 4),
            nn.GELU(),
            nn.Linear(in_dim // 4, 1),
        )
初始化函数
        if isinstance(m, nn.Linear):
            nn.init.xavier_uniform_(m.weight)
            nn.init.constant_(m.bias, 0)
```
forward函数也很简单
```
    def forward(self, x):
        out = self.layer(x)
        return out
```
我debug了一遍没注意到有啥问题，然后发现单卡是不报错的，多卡分布式训练后，就会报错定义的3个Linear层，weight和bias未收到权重。

至于您所提到的是否是有些操作会中断grad，这个改如何定位呢？

**Comment by xiuqhou** [comment]
单卡没有梯度检查过程，所以不会报错，但代码依然是有问题的，因为您定义的层的参数没有接收到梯度，也没有进行更新。

想问一下您修改代码后是否有增加新的loss? Salience-DETR会返回一个loss字典，每个key对应一个loss值的value，如果您有向这个loss字典中增加了新的loss，则需要对应地在weight_dict中设置权重。否则新增的loss就没有用到，导致相应的层没有梯度。

`detach()`和`with torch.no_grad()`会中断梯度，可以看一下您返回的out在后续计算中是否参与了类似的操作。

这个代码片段看不出来什么问题，希望您可以放更详细点的代码。如果不方便在这里公开讨论，也可以直接发给我的邮箱（xiuqhou@stu.xjtu.edu.cn），我可以帮您看看，并且不外传。

**Comment by autoencoder-zz** [comment]
> 单卡没有梯度检查过程，所以不会报错，但代码依然是有问题的，因为您定义的层的参数没有接收到梯度，也没有进行更新。
> 
> 想问一下您修改代码后是否有增加新的loss? Salience-DETR会返回一个loss字典，每个key对应一个loss值的value，如果您有向这个loss字典中增加了新的loss，则需要对应地在weight_dict中设置权重。否则新增的loss就没有用到，导致相应的层没有梯度。
> 
> `detach()`和`with torch.no_grad()`会中断梯度，可以看一下您返回的out在后续计算中是否参与了类似的操作。
> 
> 这个代码片段看不出来什么问题，希望您可以放更详细点的代码。如果不方便在这里公开讨论，也可以直接发给我的邮箱（[xiuqhou@stu.xjtu.edu.cn](mailto:xiuqhou@stu.xjtu.edu.cn)），我可以帮您看看，并且不外传。

我debug的时候看到了最后loss的计算，我没有新加loss，只是改变了原有SalienceCriterion中confidence的计算方法。

我这就整理一下代码发过去，太感谢您了
